Sources/Purchasing/Purchases/PurchasesOrchestrator.swift

```diff
@@ -551,100 +564,141 @@ final class PurchasesOrchestrator {
+            return (transaction, customerInfo, userCancelled)
+        } catch {
+            return try await self.handleSK2ProductPurchaseError(error,
+                                                                startTime: startTime,
+                                                                productId: sk2Product.id,
+                                                                promotionalOfferId: promotionalOffer?.identifier,
+                                                                winBackOfferApplied: winBackOfferApplied)
         }
+    }
 
-        let customerInfo: CustomerInfo
-
-        if let transaction = transaction {
-            customerInfo = try await self.handlePurchasedTransaction(transaction, .purchase, metadata)
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    private func handleSK2ProductPurchaseError(
+        _ error: Error,
+        startTime: Date,
+        productId: String,
+        promotionalOfferId: String?,
+        winBackOfferApplied: Bool
+    ) async throws -> PurchaseResultData {
+
+        if case StoreKitError.userCancelled = error {
+            guard !self.systemInfo.dangerousSettings.customEntitlementComputation else {
+                throw ErrorUtils.purchaseCancelledError()
+            }
 
-            self.postPaywallEventsIfNeeded()
+            self.trackPurchaseEventIfNeeded(startTime,
+                                            successful: false,
+                                            productId: productId,
+                                            promotionalOfferId: promotionalOfferId,
+                                            winBackOfferApplied: winBackOfferApplied,
+                                            storeKitVersion: .storeKit2,
+                                            purchaseResult: .userCancelled,
+                                            error: StoreKitError.userCancelled.asPublicError)
+
+            let customerInfo = try await self.customerInfoManager.customerInfo(appUserID: self.appUserID,
+                                                                               fetchPolicy: .cachedOrFetched)
+            return (transaction: nil, customerInfo: customerInfo, userCancelled: true)
         } else {
-            // `transaction` would be `nil` for `Product.PurchaseResult.pending` and
-            // `Product.PurchaseResult.userCancelled`.
-            customerInfo = try await self.customerInfoManager.customerInfo(appUserID: self.appUserID,
-                                                                           fetchPolicy: .cachedOrFetched)
-        }
+            guard !self.systemInfo.dangerousSettings.customEntitlementComputation else {
+                throw error
+            }
 
-        if !userCancelled {
-            self.trackPurchaseEventIfNeeded(storeKitVersion: .storeKit2, error: nil)
-        }
+            let purchasesError: PurchasesError = switch error {
```

Comment:
I believe this syntax (switch as an expression) is not supported in older xcode versions so we had to revert usages of it for now... We can probably use some `#if swift...` macros but might be better to write it without using those.

Reply:
Right! This is a feature in Swift 5.9 and Xcode 15 (minimum required to upload apps to the App Store) already comes with Swift 5.9. That said, I think it's possible for someone to still be using an older version of Swift even, so better to be safe than sorry. 
Thanks for the catch @tonidero!